Add serialization logic to pytree types

angelayi opened the pull request:
# What does this PR do?

[`torch.export`](https://pytorch.org/docs/main/export.html) produces a graph representation of the program which contains a pytree `TreeSpec` specifying how to flatten the inputs from whatever structure they were in before to a flattened list of inputs to pass to the graph, and unflatten the outputs from the graph to the original structure that eager mode PyTorch produces. We would like to serialize the artifact from `torch.export` and later pass it to a runtime. This means that we need a good BC/FC surface.

The pytree `TreeSpec` contains a structure telling us how the original input/output looked like. For example, if we had an input containing a tuple of a class `InputClass`, the `TreeSpec` would look something like: `TreeSpec(tuple, None, [TreeSpec[InputClass, None, []], TreeSpec[InputClass, None, []]])`. We want to serialize this into a json string consisting of the serialized type, serialized context, and recursively serialized children. Since we want to convert to a json format, only json types are allowed (strings, ints...). Therefore in this PR you can see logic to serialize a python type to a fully qualified string name.

Note that this field only exists starting from PyTorch version 2.2.

Here's an example:
```
import torch
import io
from transformers import AlbertForMaskedLM
from transformers import file_utils

model_cls = AlbertForMaskedLM
model_config = model_cls.config_class()
model = model_cls(model_config)
model.to("cuda", torch.float32)

input_dict = {"input_ids": torch.randint(0, 30000, (1, 512), device="cuda", dtype=torch.int64, requires_grad=False)}

ep = torch.export.export(model, (), input_dict)

print(ep.module_call_graph[0].signature.in_spec)
# TreeSpec(tuple, None, [TreeSpec(tuple, None, []), TreeSpec(dict, ['input_ids'], [*])])
print(ep.module_call_graph[0].signature.out_spec)
# TreeSpec(MaskedLMOutput, (<class 'transformers.modeling_outputs.MaskedLMOutput'>, ['logits']), [*])

buffer = io.BytesIO()
torch.export.save(ep, buffer)
buffer.seek(0)
loaded_ep = torch.export.load(buffer)

print(loaded_ep.module_call_graph[0].signature.in_spec)
# TreeSpec(tuple, None, [TreeSpec(tuple, None, []), TreeSpec(dict, ['input_ids'], [*])])
print(loaded_ep.module_call_graph[0].signature.out_spec)
# TreeSpec(MaskedLMOutput, (<class 'transformers.modeling_outputs.MaskedLMOutput'>, ['logits']), [*])

input_dict = {"input_ids": torch.randint(0, 30000, (1, 512), device="cuda", dtype=torch.int64, requires_grad=False)}
assert(torch.allclose(model(**input_dict).logits, loaded_ep(**input_dict).logits))
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ydshieh who seems to have reviewed pytree-related stuff before :)

angelayi committed:
Add serialized type name to pytrees

angelayi commented:
@ydshieh Sorry for the delay! The PR context changed a little, but I added more information to the PR summary, and added a test. Please let me know if you need any more clarifications!

ydshieh commented:
Hi @angelayi 

Thanks for the update. I would still love to see a code snippet (that we can run easily) in action that address the following

> Could you show us an example that we can see the serialized name being changed (without the change in this PR)?

🙏

angelayi commented:
Hi @ydshieh, I updated the PR so that comment is no longer really relevant!

angelayi commented:
bump @ydshieh

ydshieh commented:
Hi @angelayi .

Although I get the idea overall, and it LGTM, I still have a few questions and comments (like previously) and would love to see more detailed explanations.

- I still don't get why this change is `necessary`:
  - without this change, would we get some BC/FC issue, say a graph created with old torch version could not be used with newer torch version?
  - if this is the case, could you elaborate what the situation(s) is/are?
- And since this (the code content been added in this PR) is not something (very) public that a lot of users will use (it is more internal), I personally would appreciate a lot if an example that uses `torch.export` (I assume this is more public) that you mentioned in the PR description to demonstrate the situation (instead of just using `pytree.treespec_dumps` - it's fine to use this in the testing, but would help a lot to put an example in the PR description using `torch.export`).


Looking forward to your comments !

angelayi commented:
@ydshieh 

> I still don't get why this change is necessary:

Without this change, we cannot serialize exported programs for models with outputs of type `ModelOutput`, aka most HuggingFace models. We would like for users to be able to save their exported programs and load it in another time.

> I personally would appreciate a lot if an example that uses torch.export

I added an example to the PR description which exports AlbertMaskedLM, saves, and loads it. Under the hood in torch.export.save, we use pytree.treespec_dumps to save the treespec, and in torch.export.load, we use pytree.treespec_loads to load the treespec. I also linked an issue where some of our benchmarks are failing due to not having pytree serialization support.

ydshieh commented:
Thank you a lot @angelayi I will check the example ❤️ .

ydshieh commented:
review this today

ydshieh commented:
Hi! I reviewed the change along with the example you provided, I have one major question and another minor one.

Minor one: So `torch.export.save` is not working with torch < 2.2 for HF models, right. I am surprised by this fact, but I am not the one who work on this part.

### Major issue:

A variable `SERIALIZED_CLASS_TO_PYTHON_CLASS` is created and used. It works well with your script as both saving and loading are done in the same python process (which uses a buffer).

However, despite I don't have much experience on this topic, I think a common scenario/use case is: A user saves the exported program to a file. And later on, they will load the file, but in another python process. 

So the 2 scripts (saving/loading) will give `KeyError: 'transformers.modeling_outputs.MaskedLMOutput'` (see the full log at the end).

**Question**: Do you have any idea how to address this common use case?

-------------------------------------------------------------------------------------------------------------------------

#### saving
```python
import torch
import io
from transformers import AlbertForMaskedLM
from transformers import file_utils

model_cls = AlbertForMaskedLM
model_config = model_cls.config_class()
model = model_cls(model_config)
model.to("cuda", torch.float32)

input_dict = {"input_ids": torch.randint(0, 30000, (1, 512), device="cuda", dtype=torch.int64, requires_grad=False)}
ep = torch.export.export(model, (), input_dict)

fn = 'my_exported'
torch.export.save(ep, fn)
```

#### loading
```python
import torch
import io
from transformers import AlbertForMaskedLM
from transformers import file_utils

model_cls = AlbertForMaskedLM
model_config = model_cls.config_class()
model = model_cls(model_config)
model.to("cuda", torch.float32)

input_dict = {"input_ids": torch.randint(0, 30000, (1, 512), device="cuda", dtype=torch.int64, requires_grad=False)}

fn = 'my_exported'
loaded_ep = torch.export.load(fn)


input_dict = {"input_ids": torch.randint(0, 30000, (1, 512), device="cuda", dtype=torch.int64, requires_grad=False)}
assert(torch.allclose(model(**input_dict).logits, loaded_ep(**input_dict).logits))
```

#### error log

```bash
Traceback (most recent call last):
  File "temp3.py", line 14, in <module>
    loaded_ep = torch.export.load(buffer)
  File "/usr/local/lib/python3.8/dist-packages/torch/export/__init__.py", line 582, in load
    return load(
  File "/usr/local/lib/python3.8/dist-packages/torch/_export/__init__.py", line 1079, in load
    ep = deserialize(artifact)
  File "/usr/local/lib/python3.8/dist-packages/torch/_export/serde/serialize.py", line 1714, in deserialize
    ExportedProgramDeserializer(expected_opset_version)
  File "/usr/local/lib/python3.8/dist-packages/torch/_export/serde/serialize.py", line 1565, in deserialize
    GraphModuleDeserializer()
  File "/usr/local/lib/python3.8/dist-packages/torch/_export/serde/serialize.py", line 1275, in deserialize
    module_call_graph = self.deserialize_module_call_graph(serialized_graph_module.module_call_graph)
  File "/usr/local/lib/python3.8/dist-packages/torch/_export/serde/serialize.py", line 1513, in deserialize_module_call_graph
    return [
  File "/usr/local/lib/python3.8/dist-packages/torch/_export/serde/serialize.py", line 1516, in <listcomp>
    signature=self.deserialize_module_call_signature(entry.signature) if entry.signature else None,
  File "/usr/local/lib/python3.8/dist-packages/torch/_export/serde/serialize.py", line 1509, in deserialize_module_call_signature
    out_spec=treespec_loads(module_call_signature.out_spec),
  File "/usr/local/lib/python3.8/dist-packages/torch/utils/_pytree.py", line 991, in treespec_loads
    return _SUPPORTED_PROTOCOLS[protocol].json_to_treespec(json_schema)
  File "/usr/local/lib/python3.8/dist-packages/torch/utils/_pytree.py", line 953, in _json_to_treespec
    context = serialize_node_def.from_dumpable_context(json_schema["context"])
  File "/transformers/src/transformers/utils/generic.py", line 465, in _model_output_from_dumpable_context
    python_class = SERIALIZED_CLASS_TO_PYTHON_CLASS[serialized_class]
KeyError: 'transformers.modeling_outputs.MaskedLMOutput'
```

angelayi committed:
Modify context

angelayi commented:
@ydshieh 

Thanks for the detailed review!

> So there is no way to make torch.export.save work with torch < 2.2 for HF models?

No sorry, torch.export.save won't work with torch < 2.2 for HF models. But also, torch.export.save was introduced in 2.1.

> A user saves the exported program to a file. And later on, they will load the file, but in another python process.

Thanks for catching this! I made a minor change to how the flattening/unflattening is implemented, and this issue should no longer occur. Please let me know what you think!

ydshieh commented:
Hi @angelayi 

Thanks a lot! The save/load works with the latest change. But the last line

> assert(torch.allclose(model(**input_dict).logits, loaded_ep(**input_dict).logits))

in `loading` script fails as the diff is `> 1.0`

(I am using this block at the end)
```python
diff = model(**input_dict).logits - loaded_ep(**input_dict).logits

diff = torch.amax(torch.abs(diff))
print(diff)
```

However, if we put the saving/loading logic in the same script, the diff is `0.0`.

I am feeling something very strange happens.

ydshieh commented:
sorry, it's my bad!  Let me check again.

ydshieh commented on src/transformers/utils/generic.py:
```diff
@@ -306,11 +308,19 @@ def __init_subclass__(cls) -> None:
+                _torch_pytree._register_pytree_node(
+                    cls,
+                    _model_output_flatten,
+                    partial(_model_output_unflatten, output_type=cls),
```
My only left (minor) question is: would this introduce any backward compatibility issue (for a user using torch < 2.2 but before and after this PR). 

I think it is `NO` as this is really only in the scope of a python runtime: i.e. there is no saving using code before this PR and loading using code of this PR. (as saving is not working anyway before this PR or torch < 2.2).

So there shouldn't be any compatibility issue. Do you have any comment regarding this?

ydshieh reviewed (approved):
Thank you for the work and patience!

Checked with the provided code snippet: confirm that without this PR, the saving/loading of exported torch models would fail in torch 2.2.

The changes made in this PR make sense.

angelayi commented:
thanks so much for the detailed reviews!!

angelayi commented:
@amyeroberts do you when we could get this merged into the repo?

amyeroberts commented on tests/utils/test_model_output.py:
I would be great to add a test which is essentially the logic in the PR description: saving and loading a model and confirming that the logits are close. 

@ydshieh Where would be the best place for this? It could be a slow test in modeling_utils which we only run if torch 2.2 is present (is this the case for any of our nightly runs?)

ydshieh replied:
It could be in `tests/utils/test_model_output.py` (the file here) as it is to test the functionality related to `ModelOutput`.

> which we only run if torch 2.2 is present (is this the case for any of our nightly runs?)
At this moment, nightly runs having 2.3 already, so yes, it will run the new suggested test. (torch 2.2 official release is soon I believe)

amyeroberts reviewed (commented):
Thanks for adding and iterating on this! 

In particular, thank you for the detail and example in the PR description, they made it a lot easier to follow this.

angelayi commented on src/transformers/utils/generic.py:
```diff
@@ -306,11 +308,19 @@ def __init_subclass__(cls) -> None:
-                _model_output_flatten,
-                _model_output_unflatten,
-            )
+            if version.parse(get_torch_version()) >= version.parse("2.2"):
```
It resulted in a circular import error for me:

``
Traceback (most recent call last):
  File "/data/users/angelayi/transformers/tests/utils/test_model_output.py", line 21, in <module>
    from transformers import AlbertForMaskedLM, file_utils
  File "/home/angelayi/.conda/envs/transformers/lib/python3.10/site-packages/transformers/__init__.py", line 26, in <module>
    from . import dependency_versions_check
  File "/home/angelayi/.conda/envs/transformers/lib/python3.10/site-packages/transformers/dependency_versions_check.py", line 16, in <module>
    from .utils.versions import require_version, require_version_core
  File "/home/angelayi/.conda/envs/transformers/lib/python3.10/site-packages/transformers/utils/__init__.py", line 32, in <module>
    from .generic import (
  File "/home/angelayi/.conda/envs/transformers/lib/python3.10/site-packages/transformers/utils/generic.py", line 32, in <module>
    from transformers.pytorch_utils import is_torch_greater_or_equal_than_2_2
  File "/home/angelayi/.conda/envs/transformers/lib/python3.10/site-packages/transformers/pytorch_utils.py", line 22, in <module>
    from .utils import is_torch_tpu_available, logging
ImportError: cannot import name 'is_torch_tpu_available' from partially initialized module 'transformers.utils' (most likely due to a circular import) (/home/angelayi/.conda/envs/transformers/lib/python3.10/site-packages/transformers/utils/__init__.py)
```

amyeroberts replied:
OK, let's leave the check as-is then

angelayi committed:
add serde test

angelayi commented:
@amyeroberts I should've addressed the issues you mentioned, except I'm failing the formatting CI job despite running `make fixup` locally 😅 I was wondering if you know how I could fix this?

amyeroberts reviewed (commented):
Thanks for iterating! Just some small nits and then I think we're ready to merge 🤗

amyeroberts commented:
@angelayi To apply the quality changes you can run `make fixup` and push the changes applies

angelayi commented:
@amyeroberts ah it seemed to work now! thanks!

amyeroberts reviewed (approved):
Thanks for this contribution and iterating on it!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27871). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
